### DISCUSSION TITLE
I started a new tread :)

### DISCUSSION THREAD

**Discussion by stama182** [AmigaPorts/ACE#219] [Q&A]
Hello again! Thanks for all help so far.
Now winuae starts when i press F5, but I see no traces of my program in it. All commands I type to AMIGaDos fail, and I see no trace of the game.log that Im supposed to write to.  :)

So If you have even more patience, Ill be glad for some more help...

Have a nice weekend regardless!

[GoodOldWorld.zip](https://github.com/AmigaPorts/ACE/files/14960331/GoodOldWorld.zip)

**Comment by tehKaiN** [comment]
Ok, so I've added the `.vscode/launch.json` file like explained in the docs, looks like this:

```json
{
	"version": "0.2.0",
	"configurations": [
		{
			"type": "amiga",
			"request": "launch",
			"name": "Amiga 500",
			"config": "A500",
			"program": "${workspaceFolder}/build/alloldworld",
			"kickstart": "D:/usr/Documents/Amiga files/WinUAE/kick13.rom",
			"internalConsoleOptions": "openOnSessionStart"
		},
		{
			"type": "amiga",
			"request": "launch",
			"name": "Amiga 1200",
			"config": "A1200",
			"program": "${workspaceFolder}/build/alloldworld",
			"kickstart": "D:/usr/Documents/Amiga files/WinUAE/kick31.rom",
			"internalConsoleOptions": "openOnSessionStart"
		}
	]
}
```

and I get no error, executable works! Indeed, log file isn't created, but I created a breakpoint on logWrite in `genericCreate()` and execution stopped properly, so there's  that. Since path leads to alloldworld, it needs both alloldworld.exe and alloldworld.elf in that folder. That's how vscode extension works and I can't really do anything about that - sorry!

Now, about logging - you need to put ACE into debug mode. Depending how you call cmake, you can do it as follows:

- if you call it from command line (e.g. `cd build; cmake ..`), you can add `-DACE_DEBUG=ON` to the command line and it'll use that
- if you just let vscode do its trick, you can go to cmake ui, find ACE_DEBUG variable, toggle it on and hit SAVE button

![obraz](https://github.com/AmigaPorts/ACE/assets/16180344/ef86eb9b-4072-4192-a3e8-9287279e2182)

- or, if you can't be bothered with that, add `set(ACE_DEBUG ON)` in your cmakelists.txt file somewhere before the `add_subdirectory(deps/ace ace)` command.

This will enable debug mode, and now you have 2 extra choices:
- if you want to have file logging which is horribly slow, but works even outside emulators, use the `GENERIC_MAIN_LOG_PATH` define as you have it in main.c
- if you want to have fast logging directly to vscode debug console, also set the ACE_DEBUG_UAE cmake var.

Hope this helps!

**Reply by stama182** [reply]
Thank you! I did now managed to get it to stop at a debug point! I didnt get it to write to a logfile though, but I left that for later. I kept reading the tutorial, and even managed to make a program that didnt exit before I pressed escape.  However, when I tried to load a png-file I realised I hadnt compiled the toolset. Thats probably why I never saw the logs either. When I try to compile the toolset it says it can only be compiled on an amiga. I think this is it for tonight :)

**Reply by tehKaiN** [reply]
uhhhh, the toolset is meant to be built on a PC, as PC executables. Either open ace/tools folder in vscode (as in file -> open folder, so that it shows as the main folder in files panel) and built as typical cmake project (selecting PC compiler, not Bartman), or do it manually with commands from the tutorial.

Keep up the good fight, once you set it up properly it'll be smooth sailing. :)

**Reply by stama182** [reply]
Thank you for your answer. I did however already know that the toolset should be compiled on PC. This is however what I get, (On my pc.)[]
![problem](https://github.com/AmigaPorts/ACE/assets/10742883/3c10fc6e-b77b-4f09-bb44-4a65e91a9789)
(url)

I also tried downloading [winlibs-x86_64-posix-seh-gcc-12.2.0-llvm-15.0.7-mingw-w64ucrt-10.0.0-r4](https://github.com/brechtsanders/winlibs_mingw/releases/download/12.2.0-15.0.7-10.0.0-ucrt-r4/winlibs-x86_64-posix-seh-gcc-12.2.0-llvm-15.0.7-mingw-w64ucrt-10.0.0-r4.7z)) again This is what you recommend for the toolchain in the docs.

[GoodOldWorld (2).zip](https://github.com/AmigaPorts/ACE/files/14970433/GoodOldWorld.2.zip)

I admit that out of desperation I tried to edit C:\GoodOldWorld\deps\ace\tools\Cmakecache by hand, but at least it didnt make stuff worse.

:)

**Reply by tehKaiN** [reply]
This doesn't look correct - it kinda looks like it accessed cmakelists from `deps/ace` instead of `deps/ace/tools` - can you confirm that your local cmakelists.txt from tools directory looks like this? https://github.com/AmigaPorts/ACE/blob/master/tools/CMakeLists.txt

**Reply by stama182** [reply]
Yes, I can confirm they look the same. You have my entire project in the GoodOldWorld (2).zip file.

**Reply by tehKaiN** [reply]
huh, perhaps try removing following files from deps/ace/tools directory, maybe they mess things up:

- cmakecache.txt
- makefile
- cmake_install.cmake
- cmakefiles directory

then try again building tools from the deps/ace/tools/build directory.

**Reply by stama182** [reply]
![problem2](https://github.com/AmigaPorts/ACE/assets/10742883/de0981c5-b785-401b-833e-568e9557e2e1)
This is what I got after that :) As you can see I tried both "cmake build ." and 'cmake .. -G "MinGW Makefiles"' .

**Reply by tehKaiN** [reply]
the problem is that you're not doing those commands in `tools/build`, but just in `tools` directory. 

Perhaps let's try something different - a bit more automated:

- make sure the files in `tools` directory which I mentioned before don't exist
- open vscode, select `file -> open folder` and point it to `tools` directory
- hit f7, it'll ask for a compiler, so select any other except Bartman one
- it should build nicely after that

**Reply by stama182** [reply]
I finally build the tools!
![solution1](https://github.com/AmigaPorts/ACE/assets/10742883/cce8bd58-2360-41c6-8ac6-36217f450cdf)

I hade to make two things to make it work.
First add theee two lines in CMAKELists in tools folder:
set(CMAKE_C_COMPILER "D:\\mingw64\\bin\\gcc.exe")
set(CMAKE_CXX_COMPILER "D:\\mingw64\\bin\\g++.exe")

Then for some strange reason I got a builderror in build.cmake because he coudnt find windres, despite me having windres in 
powershell as 'set-alias -name windres -value D:\mingw64\bin\windres.exe'

So I replaced it with the complete path in build.cmake, and then it worked!
Thank you :)

**Reply by tehKaiN** [reply]
Nice! You can also inject those vars using `cmake .. -G "MinGW Makefiles" -DCMAKE_C_COMPILER=path/to/gcc  -DCMAKE_CXX_COMPILER=path/to/g++`. Perhaps cmake doesn't like set-alias thingy. I think it'd be also fixed with adding `D:\mingw64\bin` into PATH envvar.

**Reply by stama182** [reply]
Hi again, and hope you had a fine week.

I have now managed to load an image as a background!

Quick question:
I think the following code should manage to move the background somehow, but it doesnt. It manage to show the background,
but it doesnt move.  But it shows. So Im a little happy for that still :)

void gameGsLoop(void) {
  // This will loop forever until you "pop" or change gamestate
  // or close the game

  if(keyCheck(KEY_ESCAPE)) {
    gameExit();
  }
  else {
    
    scrollPosX = scrollPosX + 1;
    for(UWORD uwX = 0; uwX < s_pMainBuffer->uBfrBounds.uwX-64; uwX += 16) {
      for(UWORD uwY = 0; uwY < s_pMainBuffer->uBfrBounds.uwY-64; uwY += 16) {
        blitCopyAligned(pBmBackground, scrollPosX, 0, s_pMainBuffer->pBack, uwX, uwY, 16, 16);
      }
    }
   // We'll come back here later
  }
}

**Reply by tehKaiN** [reply]
`blitCopyAligned()` is an optimized variant of `blitCopy()` where the assumption is that x coords are always multiple of 16 - it doesn't do per-pixel shifting during copy. You could try using `blitCopy()` (with `MINTERM_COOKIE` as last param) instead.

However, there's an easier method for doing scrolled background. Just set up simpleBuffer with bounds bigger than the viewport (e.g. `TAG_SIMPLEBUFFER_BOUND_WIDTH, 640, TAG_SIMPLEBUFFER_BOUND_HEIGHT, 512`) and then use camera functions to move it, e.g. `cameraMoveBy(s_pMainBuffer->pCamera, bDeltaX, bDeltaY)`  or `cameraCenterAt()`. This should work flawlessly assuming you have following functions at the end of your game loop:

```c
viewProcessManagers(s_pView);
copProcessBlocks();
```

For bigger scrolls, you either need to render each frame from scratch, as you're trying to do, which will result with degraded performance (25fps tops, I think) or use tileBuffer instead of simpleBuffer, which does some wacky hardware tricks to achieve semi-infinite scroll (65k x 65k pixels).

**Reply by stama182** [reply]
Thanks again! Havent tried it yet, but this gave me hope :)

**Comment by stama182** [comment]
Yes it did work! Of course :) 
But this makes me think Im missing some documentation? There isnt much I think on the page. Or were is the main documentation about all objects in ace?
This is basicly what Ive [got:](url)
https://github.com/AmigaPorts/ACE/tree/master/docs/tools

And If you dont have anything else to do, feel free to look at my current problem :)
[game.txt](https://github.com/AmigaPorts/ACE/files/15204119/game.txt)

This code scrolls. But I would like to not scroll the lower part of the screen. 
I have defined two viewports. I made viewport1 192 high becuase I didnt want it to take up entire screen. Than I made the mainbuffer for that viewport1 bigger so that I had something to scroll. And it kinda works, but this makes viewport 1 kind of override the other viewport. I wanted somthing like lemmings, were the lower part doesnt move, as its a menu with all skills. I want viewport one to move with the scroll, and viewport 2 to always display the same image. Sort of.

**Reply by tehKaiN** [reply]
yup, the docs are a bit incomplete - still a work in progress thingy.

You almost got it right - you still need to create simplebuffer for `s_pVpSkills` because a viewport by itself doesn't have any backing bitmap - you need buffer manager for it - be it simplebuffer, scrollbuffer (for more memory-efficient scroll stuff but requiring knowledge what exactly are you doing) or the tilebuffer. If you omit it and just create plain viewport without anything else, it will most likely result with some crazy stuff on screen and/or prior viewport bleeding into the other one.
